### PR TITLE
fixes CVE-2016-6303, avoid overflow

### PR DESCRIPTION
**Pull request by yasong**
##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/mdc2/mdc2dgst.c [hunk 1]
```diff
@@ -48,7 +48,7 @@ int MDC2_Update(MDC2_CTX *c, const unsigned char *in, size_t len)
 
     i = c->num;
     if (i != 0) {
-        if (len < MDC2_BLOCK - i) {
+        if (len + i < MDC2_BLOCK) {   //repair CVE-2016-6303, avoid overflow
```
Comment (beldmit) [inline crypto/mdc2/mdc2dgst.c:51]: Could you please use C-style comments here?

File: crypto/mdc2/mdc2dgst.c [hunk 2]
```diff
@@ -48,7 +48,8 @@ int MDC2_Update(MDC2_CTX *c, const unsigned char *in, size_t len)
 
     i = c->num;
     if (i != 0) {
-        if (len < MDC2_BLOCK - i) {
+          /* repair CVE-2016-6303, avoid overflow */
```
Comment (beldmit) [inline crypto/mdc2/mdc2dgst.c:51]: Nit: comment is misaligned

**Comment by beldmit** [comment]
Could you also submit a CLA?

**Comment by kroeckx** [comment]
This reverses commit 55d83bf7c10c7b205fffa23fa7c3977491e56c07, which is the one that fixes that CVE.

**Comment by yasong** [comment]
> This reverses commit [55d83bf](https://github.com/openssl/openssl/commit/55d83bf7c10c7b205fffa23fa7c3977491e56c07), which claims to fix that CVE.

Yeah, but I see that the master is not patched

**Comment by kroeckx** [comment]
It is fixed. I pointed to the commit in master that fixed it. This just reintroduces the problem.

**Comment by yasong** [comment]
> It is fixed. I pointed to the commit in master that fixed it. This just reintroduces the problem.

I found this [mdc2dgst.c](https://github.com/openssl/openssl/blob/master/crypto/mdc2/mdc2dgst.c) was not fixed. I do not know why.

**Review by mattcaswell — CHANGES_REQUESTED** [review]
This change is clearly wrong. The original code before your PR is correct and was fixed by the commit that @kroeckx points to. I'm not sure why you are claiming that it wasn't fixed.

**Comment by t8m** [comment]
@yasong you are probably confused by what the CVE is about and about other constraints of the values in the code. The `i` can have value only between `0` and `MDC2_BLOCK - 1` (included).  So there cannot be an underflow of the unsigned size_t type of the right hand side. On the other hand the original comparison before the CVE fix could overflow on the left hand side because len is unbounded, bounded only by the size_t type range. You're reintroducing the CVE by this PR.

**Comment by yasong** [comment]
> This change is clearly wrong. The original code before your PR is correct and was fixed by the commit that @kroeckx points to. I'm not sure why you are claiming that it wasn't fixed.



> @yasong you are probably confused by what the CVE is about and about other constraints of the values in the code. The `i` can have value only between `0` and `MDC2_BLOCK - 1` (included). So there cannot be an underflow of the unsigned size_t type of the right hand side. On the other hand the original comparison before the CVE fix could overflow on the left hand side because len is unbounded, bounded only by the size_t type range. You're reintroducing the CVE by this PR.

Thanks, now I see.

**Comment by slontis** [comment]
Can this be closed?

**Comment by t8m** [comment]
Closing.

**Pull request closed without merging**